f8a_report/v2/report_generator.py

```diff
@@ -0,0 +1,488 @@
+#!/usr/bin/env python3
+# Copyright © 2020 Red Hat Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Author: Deepak Sharma <deepshar@redhat.com>
+#
+"""Daily/ Monthly Report Generator for Stack Analyses v2 API."""
+
+from report_helper import ReportHelper
+import logging
+import json
+from cve_helper import CVE
+from datetime import datetime as dt
+from v2.db_gateway import ReportQueries
+from unknown_deps_report_helper import UnknownDepsReportHelper
+from sentry_report_helper import SentryReportHelper
+from s3_helper import S3Helper
+from graph_report_generator import generate_report_for_unknown_epvs, \
+    generate_report_for_latest_version
+
+logger = logging.getLogger(__file__)
+logging.basicConfig(level=logging.INFO)
+
+
+class ReportBuilderV2(ReportHelper):
```

Comment:
Without Inheritence, there will Code Duplication. Are we Okay with that..?

Reply:
Please use composition, there is no "is-a" relationship here.

Reply:
if I understood correctly, you mean something like:

`class ReportBuilderV2(object):
    def __init__(self, ReportHelper):
        self.b = ReportHelper()`

this thing actually initialises unwanted classes. like the following:

https://github.com/fabric8-analytics/f8a-stacks-report/blob/b028168767f4afdf24f4215653bb7d748da517f4/f8a_report/report_helper.py#L46-L51

I really want to avoid Initialising any class in the constructor as this includes its own complexities: 
- In writing Tests 
- In any instance variable either we want it or not, we will end up initialising S3 and Postgres object, which I believe is not required at all. 
WDYT..?

Reply:
Yes, thats what we should be doing. Using inheritance for reuse of methods is not correct. In the inheritance case also you are initialising ReportBuilder so there is no difference there.

Reply:
there is indeed a different.. unless i call 
`super.__init()`, i am not calling the constructor of ReportHelper.
 
which in new format (composition one) we have a call to `__init__` of Report Helper.

Reply:
If the class is not instantiated, how are you able to call the methods?

Reply:
i mean in the current scenario, (Inheritance) , i'm inheriting all methods of ReportHelper but not initializing it. 
If you can see, i defined a new constructor in `ReportBuilderV2` that overrides the constructor of `ReportHelper`

So, now when class initialised, and lifecycle hook calls `__init__`, `__init__` of ReportBuilderV2 has precedence over former.
https://stackoverflow.com/questions/5166473/inheritance-and-init-method-in-python

Reply:
Ok, then why do not you declare a separate constructor in reporthelper and initialise what you want there. If you do not need anything then declare the methods that you want to use as static.

Reply:
Not aware of a class having 2 constructors. btw i will be going ahead your first suggestion: Composition one.